### PR TITLE
rtdl: Implement RELR relocations

### PR DESCRIPTION
**Pull request by Qwinci**

### CODE REVIEW

**Review by Dennisbonke — APPROVED** [review]
LGTM

File: options/rtdl/generic/linker.cpp [hunk 1]
```diff
@@ -1660,6 +1673,33 @@ void Loader::_processStaticRelocations(SharedObject *object) {
...
+
+		for(size_t i = 0; i < *relr_length; i += sizeof(ElfW(Relr))) {
+			auto entry = *reinterpret_cast<const ElfW(Relr) *>(object->baseAddress + *relr_offset + i);
+
+			// Even entry indicates the beginning address.
+			if(!(entry & 1)) {
+				addr = reinterpret_cast<ElfW(Addr) *>(object->baseAddress + entry);
+				*addr++ += object->baseAddress;
+			}else {
+				// Odd entry indicates entry is a bitmap of the subsequent locations to be relocated.
```
Comment (Geertiebear) [inline options/rtdl/generic/linker.cpp:1688]: Maybe we should ``__ensure(addr)`` here.

File: options/rtdl/generic/main.cpp [hunk 2]
```diff
@@ -106,6 +110,28 @@ extern "C" void relocateSelf() {
...
+	ElfW(Addr) *addr = nullptr;
+	for(size_t i = 0; i < relr_size; i += sizeof(ElfW(Relr))) {
+		auto entry = *reinterpret_cast<const ElfW(Relr) *>(ldso_base + relr_offset + i);
+
+		// Even entry indicates the starting address.
+		if(!(entry & 1)) {
+			addr = reinterpret_cast<ElfW(Addr) *>(ldso_base + entry);
+			*addr++ += ldso_base;
+		}else {
+			// Odd entry indicates entry is a bitmap of the subsequent locations to be relocated.
```
Comment (Geertiebear) [inline options/rtdl/generic/main.cpp:123]: Same here.

**Review by Geertiebear — APPROVED** [review]
LGTM.

**Pull request merged**